katdal/dataset.py

```diff
@@ -54,6 +54,51 @@ def array_equal(a1, a2):
         return (a1.shape == a2.shape) and np.all(a1 == a2)
 
 
+class AttrsSensors(object):
+    """Metadata in the form of attributes and sensors.
+
+    Parameters
+    ----------
+    attrs : mapping from string to object
+        Metadata attributes
+    sensors : mapping from string to :class:`SensorData` objects
+        Metadata sensor cache mapping sensor names to raw sensor data
+    name : string, optional
+        Identifier that describes the origin of the metadata (backend-specific)
+
+    """
+    def __init__(self, attrs, sensors, name='custom'):
+        self.attrs = attrs
+        self.sensors = sensors
+        self.name = name
+
+
+class VisFlagsWeights(object):
+    """Correlator data in the form of visibilities, flags and weights.
+
+    Parameters
+    ----------
+    vis : array-like of complex64, shape (*T*, *F*, *B*)
+        Complex visibility data as a function of time, frequency and baseline
+    flags : array-like of uint8, shape (*T*, *F*, *B*)
+        Flags as a function of time, frequency and baseline
+    weights : array-like of float32, shape (*T*, *F*, *B*)
+        Visibility weights as a function of time, frequency and baseline
```

Comment:
This documentation implies that these are all required, but when default-constructing a `DataSource` they are set to `None`. Is that a supported use case? Is it supported to have some of them be `None` and others not? (in which case the implementation of `shape` may need to be more robust).

Reply:
I haven't fully bedded this down yet. I was thinking of a use case where you e.g. only have vis without flags and weights, but that is unlikely in the latest data streams. It might be better to just have data = None, but then it is harder to have a unified `DataSource` that could have metadata and/or data.

Reply:
On option is to use aggregation instead of multiple inheritance, so that you have `.data.vis`, or have `.data = None` if there is no data.

Reply:
This is fixed by letting the `DataSource` construct from a `VisFlagsWeights` instead, so that vis + flags + weights are either all present or all absent.